ARCs/arc-0024.md

```diff
@@ -0,0 +1,485 @@
+  specificationVersion: string;  
+
+  /** The contract implemented by this Application **/
+  contract: Contract;
+
+  /** ARC extension 1**/
+  extensionPoint1: SomeFutureExtension1; 
+
+    ...
+
+  /** ARC extension N**/
+  extensionPoint2: SomeFutureExtensionN; 
+
+}
+```
+
+This ARC proposes, therefore, that Application Specification
+formats are versioned and subsequent ARCs increment the version number.
+
+### **4. Discovery**
+
+Consumers of an Application need to be able to find the Application Specification
+for that version of it.
+
+There are many potential 'user stories' around searching for different Applications.
+We could speculate on various needs, such as:
+
+* Search by a well known Interface name that the Application implements,
+* Search by a tag,
+* Search by a foreign language tag,
+* Search if the Application is audited and by whom,
+* By transaction fees,
+* By a community 'star' rating,
+* Etc.
+
+It is difficult to know where to draw the line between what should be community
+add-on, and what should be offered by the Algorand platform.
+
+In any case, there are some important security considerations:
+
+* The consumer needs to know that the Application author is the publisher
+of the Application Specification.
+* Whichever way the consumer obtains the Application ID on the
+Algorand blockchain, they need a mechanism to find the corresponding
+Application Specification.
+* When peripheral directories and explorers of Application
+metadata emerge, users may be exposed to phishing-type attacks where
+they are misdirected to fake application clones (eg: T1nyman, spelt
+with a 'one').
+
+At a minimum then some kind of veriable link between Author, Application and
+Specification needs to exist.
+
+This ARC proposes the following:
+
+1. Store the SHA-512/256 hash of the current Application Specification file
+in the Application itself.
+2. To save storage, declare a constant byte array, using _bytecblock_
+that is of the format "APPSPEC{HASH}" where HASH is the 32 byte
+hash of the file.
+3. The first opcode in the ApprovalProgram program is where the
+_bytecblock_ MUST be placed.
+
+To support scenarios where the Application authors want to provide an 'official'
+host for their Application Specifications, the Application itself MAY offer a URI
+to the location of the ABI spec.
+
+4. APPSPEC{HASH}{URI} where URI is the UTF-8 byte encoded URI to some locator
+for the Application's spec.
+
+To help support the scenarios where the consumer finds the Application Specification
+file first, or wants to verify version numbers, the Application Specification
+MUST include the following extension:
+
+```typescript
+interface Application {
+
+  /** The corresponding Application id  **/
+  appID: uint64;  
+
+  /** The Application version  **/
+  version: uint32;
+  
+
+}
+```
+
+### **5. Composability**
+
+It may be desirable to build Contract elements through inheriting from another Contract
+or implementing Interfaces.
```

Comment:
I would be in favor of "flattening" any such composition mechanism. So, if an app contains some state variables because some tool or HLL has created it as "inheriting" from another app or class, that need not be described here. Instead, document the field directly. The app consumer does not need to know _why_ there is a state variable, it juts needs to know that it's there, and the format of the dat in it.